File: src/services/selector_evaluation.py
"""Explicit fixture replay, never a provider execution or publication gate."""

from collections import Counter
from decimal import Decimal
import hashlib
import math
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from src.billing.routing_costs import (
    RoutingCostAggregate,
    RoutingCostObservation,
    aggregate_routing_costs,
)
from src.route_policy_contract import LLMTierSelectorPolicy
from src.router.selection.contracts import SelectorCause
from src.router.selection.parser import parse_selector_output

MAX_EVALUATION_BYTES = 262_144
MAX_EVALUATION_SAMPLES = 100
LaneId = Annotated[str, Field(pattern=r"^[a-z][a-z0-9_-]{0,31}$")]
ExactCostText = Annotated[str, Field(pattern=r"^[0-9]{1,20}(\.[0-9]{1,18})?$", max_length=39)]


class EvaluationContract(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, hide_input_in_errors=True)


class EvaluationCosts(EvaluationContract):
    selector_provider_cost: ExactCostText | None = None
    selector_customer_charge: ExactCostText | None = None
    answer_provider_cost: ExactCostText | None = None
    answer_customer_charge: ExactCostText | None = None
    baseline_answer_provider_cost: ExactCostText | None = None
    measurable_penalty: ExactCostText | None = None

    def observation(self) -> RoutingCostObservation:
        return RoutingCostObservation(
            **{
                field: Decimal(value) if value is not None else None
                for field, value in self.model_dump().items()
            }
        )


class SelectorEvaluationSample(EvaluationContract):
    expected_lane: LaneId
    prompt: str | None = Field(default=None, max_length=32768, repr=False)
    output: str | None = Field(default=None, max_length=1024, repr=False)
    failure: Literal["selector_timeout", "provider_error", "capacity_denied"] | None = None
    latency_ms: float | None = Field(default=None, ge=0, le=300_000, allow_inf_nan=False)
    answer_quality_score: float | None = Field(default=None, ge=0, le=1, allow_inf_nan=False)
    costs: EvaluationCosts = Field(default_factory=EvaluationCosts)

    @model_validator(mode="after")
    def one_outcome(self) -> "SelectorEvaluationSample":
        if (self.output is None) == (self.failure is None):
            raise ValueError("Each sample needs exactly one output or failure")
        return self


class SelectorEvaluationRequest(EvaluationContract):
    selector: LLMTierSelectorPolicy
    samples: tuple[SelectorEvaluationSample, ...] = Field(
        min_length=1, max_length=MAX_EVALUATION_SAMPLES, repr=False
    )

    @model_validator(mode="after")
    def bounded_fixture(self) -> "SelectorEvaluationRequest":
        if len(self.model_dump_json().encode("utf-8")) > MAX_EVALUATION_BYTES:
            raise ValueError("Evaluation fixtures exceed the size limit")
        allowed = {lane.id for lane in self.selector.lanes}
        if any(sample.expected_lane not in allowed for sample in self.samples):
            raise ValueError("Expected lanes must belong to this selector")
        return self


class LaneEvaluation(EvaluationContract):
    lane: str
    expected: int
    selected: int
    correct: int
    precision: float | None
    recall: float | None


class LaneConfusion(EvaluationContract):
    expected_lane: str
    selected_lane: str
    count: int


class SelectorEvaluationReport(EvaluationContract):
    basis: Literal["supplied_fixture_replay"] = "supplied_fixture_replay"
    selector_fingerprint: str
    sample_count: int
    correct_count: int
    default_count: int
    default_rate: float
    failure_counts: dict[str, int]
    lanes: tuple[LaneEvaluation, ...]
    confusion: tuple[LaneConfusion, ...]
    latency_sample_count: int
    latency_p50_ms: float | None
    latency_p95_ms: float | None
    answer_quality_sample_count: int
    answer_quality_mean: float | None
    costs: RoutingCostAggregate
    warnings: tuple[str, ...] = (
        "Fixture replay only: no provider calls, charges or publication changes.",
        "Scores, timings and costs are supplied evidence, not independently measured results.",
        "Lane agreement is not answer quality. Missing evidence remains unavailable.",
        "This report does not validate deployment eligibility for publication.",
    )


def evaluate_selector(request: SelectorEvaluationRequest) -> SelectorEvaluationReport:
    """One bounded deterministic pass with the runtime's strict output parser."""
    confusion: Counter[tuple[str, str]] = Counter()
    failures: Counter[str] = Counter()
    latencies: list[float] = []
    quality: list[float] = []
    for sample in request.samples:
        outcome = (
            SelectorCause(sample.failure)
            if sample.failure is not None
            else parse_selector_output(sample.output, request.selector)
        )
        if isinstance(outcome, SelectorCause):
            selected = request.selector.default_lane
            failures[outcome.value] += 1
        else:
            selected = outcome.id
        confusion[(sample.expected_lane, selected)] += 1
        if sample.latency_ms is not None:
            latencies.append(sample.latency_ms)
        if sample.answer_quality_score is not None:
            quality.append(sample.answer_quality_score)
    lanes = []
    for lane in request.selector.lanes:
        expected = sum(count for (label, _), count in confusion.items() if label == lane.id)
        selected = sum(count for (_, label), count in confusion.items() if label == lane.id)
        correct = confusion[(lane.id, lane.id)]
        lanes.append(
            LaneEvaluation(
                lane=lane.id,
                expected=expected,
                selected=selected,
                correct=correct,
                precision=correct / selected if selected else None,
                recall=correct / expected if expected else None,
            )
        )
    latencies.sort()
    return SelectorEvaluationReport(
        selector_fingerprint=hashlib.sha256(
            request.selector.model_dump_json().encode()
        ).hexdigest(),
        sample_count=len(request.samples),
        correct_count=sum(lane.correct for lane in lanes),
        default_count=sum(failures.values()),
        default_rate=sum(failures.values()) / len(request.samples),
        failure_counts=dict(sorted(failures.items())),
        lanes=tuple(lanes),
        confusion=tuple(
            LaneConfusion(expected_lane=a, selected_lane=b, count=count)
            for (a, b), count in sorted(confusion.items())
        ),
        latency_sample_count=len(latencies),
        latency_p50_ms=latencies[math.ceil(len(latencies) * 0.5) - 1] if latencies else None,
        latency_p95_ms=latencies[math.ceil(len(latencies) * 0.95) - 1] if latencies else None,
        answer_quality_sample_count=len(quality),
        answer_quality_mean=sum(quality) / len(quality) if quality else None,
        costs=aggregate_routing_costs(
            tuple(sample.costs.observation() for sample in request.samples)
        ),
    )

File: ui/src/lib/selectorEvaluation.ts
import type { EvaluationCostField, SelectorEvaluationRequest, SelectorEvaluationSample } from './api/selectorReports';
import { readGuidedSelector } from './routeGroupSelector';

function record(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

const COST_FIELDS: EvaluationCostField[] = ['selector_provider_cost', 'selector_customer_charge',
  'answer_provider_cost', 'answer_customer_charge', 'baseline_answer_provider_cost', 'measurable_penalty'];

function sampleFromJson(value: unknown): SelectorEvaluationSample {
  if (!record(value) || typeof value.expected_lane !== 'string') throw new Error('Each sample needs an expected_lane.');
  const allowed = ['expected_lane', 'prompt', 'output', 'failure', 'latency_ms', 'answer_quality_score', 'costs'];
  if (Object.keys(value).some((key) => !allowed.includes(key))) throw new Error('Unsupported sample field.');
  const sample: SelectorEvaluationSample = { expected_lane: value.expected_lane };
  for (const key of ['output', 'prompt'] as const) {
    if (value[key] != null && typeof value[key] !== 'string') throw new Error('Prompt and output must be text.');
    if (typeof value[key] === 'string') sample[key] = value[key];
  }
  if (value.failure != null) {
    if (value.failure !== 'selector_timeout' && value.failure !== 'provider_error' && value.failure !== 'capacity_denied') throw new Error('Unsupported failure cause.');
    sample.failure = value.failure;
  }
  if ((sample.output == null) === (sample.failure == null)) throw new Error('Each sample needs exactly one output or failure.');
  for (const key of ['latency_ms', 'answer_quality_score'] as const) {
    if (value[key] != null) {
      if (typeof value[key] !== 'number' || !Number.isFinite(value[key]) || value[key] < 0) throw new Error('Scores and timings must be non-negative finite numbers.');
      sample[key] = value[key];
    }
  }
  if (value.costs != null) {
    if (!record(value.costs)) throw new Error('Costs must be an object of exact decimal strings.');
    const costs: SelectorEvaluationSample['costs'] = {};
    for (const [key, amount] of Object.entries(value.costs)) {
      if (!COST_FIELDS.includes(key as EvaluationCostField) || (amount !== null
        && (typeof amount !== 'string' || !/^\d+(\.\d{1,18})?$/.test(amount)))) {
        throw new Error('Costs must use supported fields and non-negative exact decimal strings, or null for unknown.');
      }
      costs[key as EvaluationCostField] = amount;
    }
    sample.costs = costs;
  }
  return sample;
}

export function evaluationRequest(policy: Record<string, unknown>, raw: string): SelectorEvaluationRequest {
  if (new TextEncoder().encode(raw).byteLength > 262144) throw new Error('Samples exceed 256 KiB.');
  let values: unknown;
  try { values = JSON.parse(raw); } catch { throw new Error('Enter a JSON array of labeled samples.'); }
  if (!Array.isArray(values) || values.length < 1 || values.length > 100) throw new Error('Use between 1 and 100 samples.');
  const selector = readGuidedSelector(policy);
  if (selector.state !== 'enabled') throw new Error('Configure a supported selector first.');
  const request: SelectorEvaluationRequest = {
    selector: {
      kind: 'llm-tier', classifier_deployment_id: selector.classifier,
      default_lane: selector.defaultLane, timeout_ms: Number(selector.timeoutMs),
      max_input_chars: Number(selector.maxInputChars),
      lanes: selector.lanes.map((lane) => ({ ...lane, rank: Number(lane.rank) })),
    },
    samples: values.map(sampleFromJson),
  };
  if (new TextEncoder().encode(JSON.stringify(request)).byteLength > 262144) throw new Error('Evaluation exceeds 256 KiB.');
  return request;
}
